File: addTodaysNYT.py
from urllib import request
import json
import html
import calendar

BASE_URL = 'http://localhost:3021'
DATE = ""
def fetch_puzzle(date="current"):
    # 'https://www.xwordinfo.com/JSON/Data.ashx?date=2/12/2025'
    headers = {
        "Referer":"https://www.xwordinfo.com/",
    }
    request_params = request.Request(
        url="https://www.xwordinfo.com/JSON/Data.ashx?date={}".format(date), 
        headers=headers
    )
    req = request.urlopen(request_params).read()
    return json.loads(req)

def request_pid():
    PID_URL = '{}/api/counters/pid'.format(BASE_URL)
    req =  request.Request(PID_URL, method='POST')
    resp = request.urlopen(req).read()
    pidresp = json.loads(resp)
    return pidresp['pid']

def xword_to_dfac(obj):
    grid = []
    circles = []
    shades = [] # TODO
    row_size = obj.get('size').get('rows')
    col_size = obj.get('size').get('cols')
    for i in range(0, row_size):
        row = []
        grid.append(row)
        for j in range(0, col_size):
            row.append(obj.get('grid')[(i*col_size)+j])
            if obj.get('circles') and obj.get('circles')[(i*col_size)+j]:
                circles.append(i*col_size + j)
    game_type = 'Daily Puzzle' if len(grid) > 10 else 'Mini Puzzle'
    title = obj.get('title', '')
    if 'Times' not in title:
        mon,day,yr = obj.get('date').split('/')
        title = '{} NY Times, {}, {} {}, {}'.format(title, obj.get('dow', ''), calendar.month_name[int(mon)], day, yr)
    author = obj.get('author', '')
    if obj.get('editor'):
        author = author + ' / {}'.format(obj['editor'])
    description = obj.get('description', '')

    def parseClueNum(clue):
        buf = ''
        for i in range(0, len(clue)):
            if clue[i] == '.':
                break
            buf += clue[i]
        return int(buf)

    across_size = parseClueNum(obj.get('clues').get('across')[-1])
    down_size = parseClueNum(obj.get('clues').get('down')[-1])

    clues = {
        "across": [None] * (across_size + 1),
        "down": [None] * (down_size + 1)
    }

    for direction in ['across', 'down']:
        for clue in obj.get('clues').get(direction):
            idx = parseClueNum(clue)
            offset = len(str(idx))+2
            clues[direction][idx] = html.unescape(clue[offset:])

    return {
      "grid": grid,
      "circles": circles,
      "shades": shades,
      "info": {
        "type": game_type,
        "title": title,
        "author": author,
        "description": description,
      },
      "clues": clues
    }

def post_puzzle(puzzle, pid, is_public=True):
    PUZZLE_URL = '{}/api/puzzle'.format(BASE_URL)
    data = {
        'puzzle': puzzle,
        'pid': pid,
        'isPublic': is_public,
    }
    req = request.Request(PUZZLE_URL, data=json.dumps(data).encode(), method='POST')
    req.add_header('Content-Type', 'application/json')
    resp = request.urlopen(req).read()
    return json.loads(resp)

def main():
    xword_puzzle = fetch_puzzle(DATE)
    pid = request_pid()
    dfac_puzzle = xword_to_dfac(xword_puzzle)
    print(post_puzzle(dfac_puzzle, pid, True))

if __name__ == '__main__':
    main()

File: ts/src/addTodaysNYT.ts
import { JsonResponse, DFACpuzzle } from './interfaces';

const BASE_URL = 'http://localhost:3021';

async function getData(): Promise<JsonResponse> {
  let response = await fetch('https://www.xwordinfo.com/JSON/Data.ashx?date=2/12/2025', {
    headers: {
      Referer: 'https://www.xwordinfo.com'
    }
  });
  let data: JsonResponse = await response.json();
  return data;
}

async function fetchPid(): Promise<string> {
  let response = await fetch(`${BASE_URL}/api/counters/pid`, {
    method: "POST"
  });
  let data: { pid: string } = await response.json();
  return data.pid;
}
async function postPuzzle(puzzle: any, pid: string, isPublic: boolean): Promise<void> {
  await fetch(`${BASE_URL}/api/puzzle`, {
    method: "POST",
    body: JSON.stringify({ puzzle: puzzle, pid, isPublic })
  });
}

function convertToDFAC(json: JsonResponse) {
    const grid: number[][] = [];
    const circles: number[] = [];
    const shades: number[] = [];
    for(let i = 0; i < json.size.rows; i++) {
      const row: number[] = [];
      grid.push(row);
      for(let j = 0; j < json.size.cols; j++) {
        row.push(
          json.grid[(i*json.size.cols)+j]
        );
        if (json.circles && json.circles[(i*json.size.cols)+j]) {
          circles.push(i * json.size.cols + j);
        }
      }
    }
    const type = grid.length > 10 ? 'Daily Puzzle' : 'Mini Puzzle';
    const title = json.title || '';
    let author = json.author || '';
    if (json.editor) author += ` / ${json.editor}`;
    const description = json.description || '';
    function parseClueNum(clue: string) {
      let buf = '';
      for(let i = 0; i < clue.length && clue[i] !== '.'; i++) {
        buf+= clue[i];
      }
      return parseInt(buf);
    }
    var decodeHtmlEntity = function(str: string) {
      str = str.replaceAll('&quot;', '"');
      return str.replace(/&#(\d+);/g, function(match, dec) {
        return String.fromCharCode(dec);
      });
    };
    const acrossSize = parseClueNum(json.clues.across[json.clues.across.length-1]);
    const downSize = parseClueNum(json.clues.down[json.clues.down.length-1]);
    const across = new Array(acrossSize);
    const down = new Array(downSize);
    for(let clue of json.clues.across) {
      const idx = parseClueNum(clue);
      const offset = String(idx).length+2;
      across[idx] = decodeHtmlEntity(clue.slice(offset));
    }
    for(let clue of json.clues.down) {
      const idx = parseClueNum(clue);
      const offset = String(idx).length+2;
      down[idx] = decodeHtmlEntity(clue.slice(offset));
    }

    const result: DFACpuzzle = {
      grid,
      circles,
      shades,
      info: {
        type,
        title,
        author,
        description,
      },
      clues: {across, down},
    };
    return result;
  }

  async function main() {
    const data = await getData();
    const converted = convertToDFAC(data);
    const pid = await fetchPid();
    postPuzzle(converted, pid, true);
  }

main();